spec/abci++/abci++_basic_concepts.md

```diff
@@ -0,0 +1,467 @@
+
+## Deterministic State-Machine Replication
+
+[&uparrow; Back to Outline](#outline)
+
+ABCI++ applications must implement deterministic finite-state machines to be
+securely replicated by the Tendermint consensus engine. This means block execution
+over the Consensus Connection must be strictly deterministic: given the same
+ordered set of transactions, all nodes will compute identical responses, for all
+successive `FinalizeBlock` calls. This is critical because the
+responses are included in the header of the next block, either via a Merkle root
+or directly, so all nodes must agree on exactly what they are.
+
+For this reason, it is recommended that application state is not exposed to any
+external user or process except via the ABCI connections to a consensus engine
+like Tendermint Core. The Application must only change its state based on input
+from block execution (`FinalizeBlock` calls), and not through
+any other kind of request. This is the only way to ensure all nodes see the same
+transactions and compute the same results.
+
+Some Applications may choose to implement immediate execution, which entails executing the blocks
+that are about to be proposed (via `PrepareProposal`), and those that the Application is asked to
+validate (via `ProcessProposal`). However, the state changes caused by processing those
+proposed blocks must never replace the previous state until `FinalizeBlock` confirms
+the block decided.
+
+Additionally, vote extensions or the validation thereof (via `ExtendVote` or
+`VerifyVoteExtension`) must *never* have side effects on the current state.
+They can only be used when their data is provided in a `RequestPrepareProposal` call.
+
+If there is some non-determinism in the state machine, consensus will eventually
+fail as nodes disagree over the correct values for the block header. The
+non-determinism must be fixed and the nodes restarted.
+
+Sources of non-determinism in applications may include:
+
+- Hardware failures
+    - Cosmic rays, overheating, etc.
+- Node-dependent state
+    - Random numbers
+    - Time
+- Underspecification
+    - Library version changes
+    - Race conditions
+    - Floating point numbers
+    - JSON or protobuf serialization
+    - Iterating through hash-tables/maps/dictionaries
+- External Sources
+    - Filesystem
+    - Network calls (eg. some external REST API service)
+
+See [#56](https://github.com/tendermint/abci/issues/56) for original discussion.
+
+Note that some methods (`Query, CheckTx, FinalizeBlock`) return
+explicitly non-deterministic data in the form of `Info` and `Log` fields. The `Log` is
+intended for the literal output from the Application's logger, while the
+`Info` is any additional info that should be returned. These are the only fields
+that are not included in block header computations, so we don't need agreement
+on them. All other fields in the `Response*` must be strictly deterministic.
+
+## Events
```

Comment:
I would say so. Events are only returned as a reaction to `FinalizeBlock` (the ones coming from `CheckTx` have been removed as they were ignored). They are typically used by clients to look up past information by querying full nodes. So it would look odd that the answer depended on the node you are querying.

Having said that, they are not hashed into the next block's header, so a non-deterministic returned value would not necessarily be detected (as happens with TxResults, AppHash, Validators, etc). I have looked it up in the "core" spec, and it explicitly says that events are not hashed into the header (just as `Log`, `Info`) [here](https://github.com/tendermint/tendermint/blob/master/spec/core/data_structures.md#header) (see description of field `LastResultHash`).

Maybe we need to further discuss this.